kiwi/builder/disk.py

```diff
@@ -554,9 +527,13 @@ def create_install_media(self, result_instance):
         :rtype: instance of :class:`Result`
         """
         if self.install_media:
+            if self.initrd_system == 'dracut':
+                boot_image = None
+            else:
+                boot_image = self.boot_image
```

Comment:
Reusing the initrd instance for kiwi custom boots makes sense as otherwise it would make the image build way longer and rebuild the same (or almost the same) initrd twice. Given the initrd creation using KIWI is suboptimal all makes sense here.

Reply:
I think I can explain the reasons that lead to this concept. In the boot image initrd ```self.boot_image``` there needs to be information which is created before the actual initrd creation process as part of the disk builder process. This information is persistently stored in self.boot_image such that it's not required to add it when the install media is created. So the reason when we need this is in builder/disk.py here

```python
# prepare for install media if requested
if self.install_media:
    log.info('Saving boot image instance to file')
    self.boot_image.dump(
        self.target_dir + '/boot_image.pickledump'
    )
```

Information that is placed in self.boot_image which we potentially use later is this:

```
if self.boot_is_crypto:
    self.boot_image.include_file(...)

...

if self.initrd_system == 'dracut':
    if self.root_filesystem_is_multipath is False:
        self.boot_image.omit_module('multipath')
    if self.root_filesystem_is_overlay:
        self.boot_image.include_module('kiwi-overlay')
        self.boot_image.write_system_config_file(
            config={'modules': ['kiwi-overlay']}
        )
    if self.disk_resize_requested:
        self.boot_image.include_module('kiwi-repart')

...

def _write_partition_id_config_to_boot_image(self):
    ...
    self.boot_image.include_file(...)

def _write_raid_config_to_boot_image(self):
    ...
    self.boot_image.include_file(...)

def _write_crypttab_to_system_image(self):
    ...
    self.boot_image.include_file(...)

def _write_generic_fstab_to_boot_image(self, device_map):
    ...
    if self.initrd_system == 'kiwi':
        self._write_generic_fstab(device_map, self.boot_image.setup)

# this one can be considered obsolete...
def _write_recovery_metadata_to_boot_image(self):
    ...
    self.boot_image.include_file()
```

So overall if we wouldn't have the need to preserve data for the install initrd creation process none of the pickle()
code would be needed.

I have not yet looked at your refactor but I hope this explains the reasons why I implemented it that way ?

Actually a better implementation allows to transport this information from the disk builder instance to the
BootImage instance such that there is a loose coupling. I implemented that coupling by creating the BootImage instance first and serialize it via object pickle such that I can re-create that instance at any later point in time. I admit this was not the best idea ;) and kiwi also has changed a lot since that time

Does this give you a better idea why it was done this way ?
If yes and you look at your refactor code, do you still think it does not create regressions on install media creation, for example on crypto images to include /etc/crypttab in the install image initrd ?

Thanks much for this effort